### PR TITLE
[6X] gpstate -e : Display ongoing recovery progress

### PR DESCRIPTION
**Pull request by kaknikhil**
Added new section under gpstate -e, which displays progress of mirrorss
in recovery

-Edited buildMirrorSegments to create new file (recovery_progress.file)
and updates the current recovery progress, deletes the file after
recoveries are completed.
-Added new section under gpstate -e to display progress of mirrors in
recovery
-Added _parse_recovery_progress_data(), which will look for the
recovery_progress.file and populate the required fields -Added behave
and unit tests for above scenario

Sample output:
-----------------------------------------------------
Segments in recovery
Segment                     Port   Recovery type   Completed bytes (kB)   Total bytes (kB)   Percentage completed
hmaddileti-a01.vmware.com   7002   full            1164848                1371715            84%
hmaddileti-a01.vmware.com   7003   incremental     1171384                1371875            85%
hmaddileti-a01.vmware.com   7004   full            116488                 1371715            60%

Co-authored-by: Divyesh Vanjare <vanjared@vmware.com>
Co-authored-by: Nikhil Kak <nkak@vmware.com>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: gpMgmt/bin/gppylib/operations/buildMirrorSegments.py [hunk 1]
```diff
@@ -387,15 +398,30 @@ def print_progress():
...
             outfile.flush()
 
-        while not self.__pool.join(interval):
-            print_progress()
-            written = True
+        written = False
+        combined_progress_filepath = get_recovery_progress_file(gplog)
+        pattern = re.compile(get_recovery_progress_pattern())
+        try:
+            with open(combined_progress_filepath, 'w') as combined_progress_file:
```
Comment (kaknikhil) [inline gpMgmt/bin/gppylib/operations/buildMirrorSegments.py:415]: There is an issue with opening the file only once not inside the `print_progress` function. Consider the case where  one of the mirrors completes the recovery much faster than all others. As a result of this open being outside the loop, the fast mirror won't go away from the `gpstate -e` view even when it completes it's recovery. This is very easily reproduced by using `-b 1 and -B 1` when running gprecoverseg.

This issue exists on master as well but we can fix this in a separate PR

**Review by pobbatihari — APPROVED** [review]

**Comment by kaknikhil** [comment]
Here is the list of scenarios we tested

**On a cluster with 3 segment hosts and 6 primaries and 6 mirrors - total 12**
*Bring down two primaries and 1 mirror*

*  -b 1  -B 2 - worked as expected (*with the bug where the mirror that finishes doesn't disappear from gpstate output*)
*  -b 1 -B 2 -s - worked as expected
*  normal full recovery  gprecoverseg -aF - worked as expected
*  normal incremental recovery  gprecoverseg -a - worked as expected
*  failure scenario with gprecoverseg -aF (make sdw2 mirror -w) - worked as expected
*  failure scenario with gprecoverseg -a (make sdw2 mirror delete pg_xlog) - worked as expected
*  failure scenario with gprecoverseg -aF (kill process on sdw2 in between) - worked as expected
*  ctrl -c after recovery starts - combined progress was not there as expected
    *  One issue we noticed here is that gpstate kept showing the output even after removing the lock file
*  pass and failure scenario with -i - not tested

**So to summarize, two issues**
1. we don’t remove from the gpstate output, the mirror that has finished its recovery or has failed
1. If gprecoverseg is issued a kill -9 and then the lock file is removed, we still keep displaying the recovery information in gpstate from the killed process
